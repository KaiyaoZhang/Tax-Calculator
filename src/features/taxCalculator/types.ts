export interface TaxBandType {
  min: number;
  max?: number;
  rate: number;
};

export interface TaxBracketsType {
  tax_brackets: TaxBandType[];
};

export interface FormFiledsType {
  income?: string;
  year?: string;
};

export interface BandTaxTableRowType {
  key: string;
  range: string;
  rate: string;
  tax: string;
}