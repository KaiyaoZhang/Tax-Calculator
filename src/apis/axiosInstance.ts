import axios, { AxiosInstance } from "axios";

// Create an Axios instance with custom configuration
const instance: AxiosInstance = axios.create({
  // Base URL for all API requests â€“ can be set through environment variables or fallback to localhost
  baseURL: process.env.REACT_APP_API_BASE_URL || "http://localhost:5001",
  // Set timeout limit (in milliseconds) for requests
  timeout: 100000,
  // Default headers for every request made using this instance
  headers: {
    "Content-Type": "application/json",
  },
});

export default instance;
