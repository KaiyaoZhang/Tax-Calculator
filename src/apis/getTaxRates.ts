import axios from "./axiosInstance";
import { TaxBracketsType } from "../features/taxCalculator/types";

export const getTaxRates = async (year: string): Promise<TaxBracketsType> => {
  try {
    const res = await axios.get<TaxBracketsType>(
      `/tax-calculator/tax-year/${year}`
    );
    return res.data;
  } catch (e) {
    throw e; // rethrow the error so can handle it in the UI
  }
};
