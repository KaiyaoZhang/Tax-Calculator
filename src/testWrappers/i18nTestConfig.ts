import i18n from "i18next";
import { initReactI18next } from "react-i18next";
import enTranslation from "../../public/locales/en/translation.json";
import frTranslation from "../../public/locales/fr/translation.json";

i18n.use(initReactI18next).init({
  lng: "en", // Default test language
  fallbackLng: "en",
  resources: {
    en: { translation: enTranslation },
    fr: { translation: frTranslation },
  },
  interpolation: { escapeValue: false },
});

export default i18n;
