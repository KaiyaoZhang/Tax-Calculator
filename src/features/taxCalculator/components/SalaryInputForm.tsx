import { useMemo, Dispatch, SetStateAction, useCallback, JSX } from "react";
import { Button, Form, Input, Select, notification } from "antd";
import type { FormProps } from "antd";
import { useTranslation } from "react-i18next";
import { getTaxRates } from "../../../apis";
import { FormFiledsType, TaxBracketsType } from "../types";
import styles from "../index.module.scss";

const { Option } = Select;

interface SalaryInputFormProps {
  setTaxBrackets: Dispatch<SetStateAction<TaxBracketsType>>;
  setAnnualIncome: Dispatch<SetStateAction<string>>;
  setSubmitForm: Dispatch<SetStateAction<boolean>>;
  submitForm: boolean;
}

const SalaryInputForm = ({
  setTaxBrackets,
  setAnnualIncome,
  setSubmitForm,
  submitForm,
}: SalaryInputFormProps): JSX.Element => {
  const [form] = Form.useForm();
  const { t } = useTranslation();
  const [api, contextHolder] = notification.useNotification();

  const selectOptions: string[] = useMemo(
    () => ["2019", "2020", "2021", "2022"],
    []
  );

 const incomeInputRules = useMemo(
   () => [
     {
       required: true,
       message: t("income_validate_empty"), // Validation: Field cannot be empty
     },
     {
       pattern: /^\d+(\.\d{1,2})?$/,
       message: t("income_validate_number"), // Validation: Must be a valid number (up to two decimal places)
     },
     {
       validator: (_: unknown, value: string) => {
         if (!value || parseFloat(value) > 0) {
           return Promise.resolve();
         }
         return Promise.reject(
           new Error(t("income_validate_greater_than_zero"))
         );
       },
       // Validation: Must be greater than 0
     },
   ],
   [t]
 );


  const yearSelectRules = useMemo(
    () => [{ required: true, message: t("year_validate_empty") }], // Validation: Tax year cannot be empty
    [t]
  );

  const onFinish: FormProps<FormFiledsType>["onFinish"] = useCallback(
    async (values: FormFiledsType) => {
      setSubmitForm(true);
      try {
        const { income, year } = values;
        if (income && year) {
          setAnnualIncome(income);
          const taxRates: TaxBracketsType | null = await getTaxRates(year);
          if (taxRates) {
            setTaxBrackets(taxRates);
          }
        }
      } catch (e) {
        //Handle the API error
        api["error"]({
          message: t("notification"),
          description: t("APIError"),
        });
        setTaxBrackets({ tax_brackets: [] });
        console.error("API Error:", e);
      } finally {
        setSubmitForm(false);
      }
    },
    [setAnnualIncome, setTaxBrackets, t]
  );

  return (
    <div className={styles.form}>
      {contextHolder}
      <Form layout={"inline"} form={form} name="basic" onFinish={onFinish}>
        <Form.Item<FormFiledsType>
          label={t("income")}
          name={"income"}
          rules={incomeInputRules}
        >
          <Input prefix="$" placeholder="0" data-testid="income_input" />
        </Form.Item>
        <Form.Item<FormFiledsType>
          label={t("year")}
          name={"year"}
          rules={yearSelectRules}
        >
          <Select
            placeholder={t("year_placeholder")}
            className={styles.formSelect}
            data-testid="year_select"
          >
            {selectOptions.map((year, index) => (
              <Option
                value={year}
                key={index}
                data-testid={`select_option_${year}`}
              >
                {year}
              </Option>
            ))}
          </Select>
        </Form.Item>
        <Form.Item>
          <Button type="primary" htmlType="submit" loading={submitForm}>
            {t("submit")}
          </Button>
        </Form.Item>
      </Form>
    </div>
  );
};

export default SalaryInputForm;
