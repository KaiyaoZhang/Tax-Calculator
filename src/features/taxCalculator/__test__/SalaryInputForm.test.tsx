import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { I18nextProvider } from "react-i18next";
import i18nTestConfig from "../../../testWrappers/i18nTestConfig";
import { TaxBracketsType } from "../types";
import { getTaxRates } from "../../../apis";
import SalaryInputForm from "../components/SalaryInputForm";

const { t } = i18nTestConfig;

jest.mock("../../../apis", () => ({
  getTaxRates: jest.fn(),
}));

afterEach(() => {
  jest.clearAllMocks();
});

const mockResponse: TaxBracketsType = {
  tax_brackets: [
    {
      max: 48535,
      min: 0,
      rate: 0.15,
    },
    {
      max: 97069,
      min: 48535,
      rate: 0.205,
    },
    {
      max: 150473,
      min: 97069,
      rate: 0.26,
    },
    {
      max: 214368,
      min: 150473,
      rate: 0.29,
    },
    {
      min: 214368,
      rate: 0.33,
    },
  ],
};

describe("Salary input iorm page", () => {
  const salaryInputFormProps = {
    setTaxBrackets: jest.fn(),
    setAnnualIncome: jest.fn(),
    setSubmitForm: jest.fn(),
    submitForm: false,
  };
  it("renders the form inputs and submit button", () => {
    const { container } = render(
      <I18nextProvider i18n={i18nTestConfig}>
        <SalaryInputForm {...salaryInputFormProps} />
      </I18nextProvider>
    );

    expect(container).toMatchSnapshot();

    expect(screen.getByText(t("submit"))).toBeInTheDocument();
    expect(screen.getByText(t("year"))).toBeInTheDocument();
    expect(screen.getByText(t("income"))).toBeInTheDocument();
  });

  it("submits the form with valid input", async () => {
    (getTaxRates as jest.Mock).mockResolvedValue(mockResponse);
    render(<SalaryInputForm {...salaryInputFormProps} />);

    fireEvent.change(screen.getByTestId("income_input"), {
      target: { value: "50000" },
    });

    fireEvent.mouseDown(screen.getByText(t("year_placeholder")));

    fireEvent.click(screen.getByTestId("select_option_2022"));

    fireEvent.click(screen.getByText(t("submit")));

    await waitFor(() => {
      expect(salaryInputFormProps.setAnnualIncome).toHaveBeenCalledWith(
        "50000"
      );
      expect(getTaxRates).toHaveBeenCalled();
    });
  });

  it("submits the form with valid input but API fails", async () => {
    const consoleErrorSpy = jest
      .spyOn(console, "error")
      .mockImplementation(() => {}); // Suppress console.error

    (getTaxRates as jest.Mock).mockRejectedValue(new Error("API Error"));

    render(<SalaryInputForm {...salaryInputFormProps} />);

    fireEvent.change(screen.getByTestId("income_input"), {
      target: { value: "50000" },
    });

    fireEvent.mouseDown(screen.getByText(t("year_placeholder")));

    fireEvent.click(screen.getByTestId("select_option_2022"));

    fireEvent.click(screen.getByText(t("submit")));

    await waitFor(() => {
      expect(salaryInputFormProps.setTaxBrackets).toHaveBeenCalledWith({
        tax_brackets: [],
      });
    });

    consoleErrorSpy.mockRestore(); // Restore console.error after the test
  });

  it('it should not sumbit the form when there are validation fails', async () => {
    render(<SalaryInputForm {...salaryInputFormProps}/>);

    fireEvent.click(screen.getByText(t("submit")));

    await waitFor(() => {
      expect(getTaxRates).not.toHaveBeenCalled();
    });
  })
});
