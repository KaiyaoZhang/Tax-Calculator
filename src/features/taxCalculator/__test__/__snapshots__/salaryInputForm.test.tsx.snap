// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Salary input iorm page renders the form inputs and submit button 1`] = `
<div>
  <div
    class="form"
  >
    <form
      class="ant-form ant-form-inline css-dev-only-do-not-override-19lec04"
      id="basic"
    >
      <div
        class="ant-form-item css-dev-only-do-not-override-19lec04"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-19lec04"
        >
          <div
            class="ant-col ant-form-item-label css-dev-only-do-not-override-19lec04"
          >
            <label
              class="ant-form-item-required"
              for="basic_income"
              title="Annual Income"
            >
              Annual Income
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-19lec04"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <span
                  class="ant-input-affix-wrapper css-dev-only-do-not-override-19lec04 ant-input-outlined"
                >
                  <span
                    class="ant-input-prefix"
                  >
                    $
                  </span>
                  <input
                    aria-required="true"
                    class="ant-input css-dev-only-do-not-override-19lec04"
                    data-testid="income_input"
                    id="basic_income"
                    placeholder="0"
                    type="text"
                    value=""
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-form-item css-dev-only-do-not-override-19lec04"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-19lec04"
        >
          <div
            class="ant-col ant-form-item-label css-dev-only-do-not-override-19lec04"
          >
            <label
              class="ant-form-item-required"
              for="basic_year"
              title="Tax Year"
            >
              Tax Year
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-19lec04"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  aria-required="true"
                  class="ant-select ant-select-outlined ant-select-in-form-item formSelect css-dev-only-do-not-override-19lec04 ant-select-single ant-select-show-arrow"
                  data-testid="year_select"
                >
                  <div
                    class="ant-select-selector"
                  >
                    <span
                      class="ant-select-selection-wrap"
                    >
                      <span
                        class="ant-select-selection-search"
                      >
                        <input
                          aria-autocomplete="list"
                          aria-controls="basic_year_list"
                          aria-expanded="false"
                          aria-haspopup="listbox"
                          aria-owns="basic_year_list"
                          aria-required="true"
                          autocomplete="off"
                          class="ant-select-selection-search-input"
                          id="basic_year"
                          readonly=""
                          role="combobox"
                          style="opacity: 0;"
                          type="search"
                          unselectable="on"
                          value=""
                        />
                      </span>
                      <span
                        class="ant-select-selection-placeholder"
                      >
                        Select tax year
                      </span>
                    </span>
                  </div>
                  <span
                    aria-hidden="true"
                    class="ant-select-arrow"
                    style="user-select: none;"
                    unselectable="on"
                  >
                    <span
                      aria-label="down"
                      class="anticon anticon-down ant-select-suffix"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-form-item css-dev-only-do-not-override-19lec04"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-19lec04"
        >
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-19lec04"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <button
                  class="ant-btn css-dev-only-do-not-override-19lec04 ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
                  type="submit"
                >
                  <span>
                    Submit
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
